examples/pipelines/datacomp/components/dedup_image_embeddings/src/main.py

```diff
@@ -0,0 +1,89 @@
+"""This component deduplicates images based on embeddings.
+
+As proposed in [Abbas et al., 2022](https://arxiv.org/abs/2303.09540).
+
+Implementation is based on https://github.com/conceptofmind/SemDeDup/tree/main.
+"""
+import logging
+
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+from scipy.spatial.distance import cdist
+from sklearn.metrics.pairwise import cosine_similarity
+
+from fondant.component import DaskTransformComponent
+
+logger = logging.getLogger(__name__)
+
+
+def sort_by_centroid_distance(embeddings, centroid, descending=True):
+    distances = cdist(embeddings, centroid.reshape(1, -1), "euclidean")
+    sorted_indices = np.argsort(distances, axis=0)
+    if descending:
+        sorted_indices = sorted_indices[::-1]
+    return embeddings[sorted_indices.flatten()]
+
+
+class DedupImageEmbeddingsComponent(DaskTransformComponent):
+    """Component that deduplicates images based on embeddings."""
+
+    def transform(self, dataframe: dd.DataFrame, epsilon: float) -> dd.DataFrame:
+        """
+        Args:
+            dataframe: Dask dataframe
+            epsilon: Epsilon to use as threshold for cosine similarity.
+
+        Returns:
+            Dask dataframe
+        """
+        num_clusters = dataframe["image_cluster_label"].unique()
+
+        indices_to_keep = []
+        for i in num_clusters:
+            mask = dataframe["image_cluster_label"] == i
+            cluster_i_embeddings = dataframe[mask]["image_embedding"]
+            cluster_embeddings = np.vstack(cluster_i_embeddings)
+
+            print(f"Shape of embeddings cluster {i}:", list(cluster_embeddings.shape))
+
+            # get cluster centroid
+            cluster_center = np.mean(cluster_embeddings, axis=0)
+
+            # sort the cluster embeddings by the distance to the cluster centroid
+            cluster_embeddings = sort_by_centroid_distance(
```

Comment:
Is the sorting inside of a single partition really needed?

Reply:
It's based on the pseudocode provided in the semdedup paper: https://arxiv.org/abs/2303.09540.

I also wonder whether it influences the results